src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/DocumentContextExtensions.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Extensions
+{
+    internal static class DocumentContextExtensions
+    {
+        internal static async Task<Projection> GetProjectionAsync(this DocumentContext documentContext, int absoluteIndex, RazorDocumentMappingService documentMappingService, CancellationToken cancellationToken)
```

Comment:
I'd throw this onto the document context object itself, suppppper useful. That way the factory can pass in the mapping service too so it doesn't need to always be passed in.

Reply:
The passing in of the service is annoying, but it feels weird to me for document contexts to hold references to services.

What about flipping this around, to be on the service? ie, `_documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken)`?

Reply:
LOVE. yes gogogogo